docs/source/howto/codes.rst

```diff
@@ -3,173 +3,200 @@
+
+In the following, our |Code| will be the `bash` executable, and our "input file" will be a `bash` script ``aiida.in`` that sums two numbers and prints the result:
+
+.. code-block:: bash
+
+   echo $(( numx + numy ))
+
+We will run this is as:
 
 .. code-block:: bash
 
     /bin/bash < aiida.in > aiida.out
 
+thus writing the sum of the two numbers ``numx`` and ``numy`` (provided by the user) to the output file ``aiida.out``.
+
+
 .. _how-to:codes:interfacing:
 
+.. todo::
+
+    Add to preceding sentence: :ref:`the communication with external machines<how-to:codes:transport>` and the interaction with its :ref:`scheduling software<how-to:codes:scheduler>`.
+
+
 Interfacing external codes
 ==========================
 
-To provide AiiDA with the set of instructions, required to run a code, one should subclass the |CalcJob| class and implement the following two key methods:
+We start our ``calcjob.py`` script by subclassing the |CalcJob| class:
+
+.. literalinclude:: ../../../aiida/calculations/arithmetic/add.py
+    :language: python
+    :lines: 10-18
+
+In the following, we will tell AiiDA how to run our code by implementing two key methods:
 
  #. :py:meth:`~aiida.engine.processes.calcjobs.calcjob.CalcJob.define`
  #. :py:meth:`~aiida.engine.processes.calcjobs.calcjob.CalcJob.prepare_for_submission`
 
-We will now show how each of these can be implemented.
+Defining the spec
+-----------------
 
-Defining the specifications
----------------------------
-
-The |define| method is where one specifies the different inputs that the caller of the |CalcJob| will have to provide in order to run the code, as well as the outputs that will be produced (exit codes will be :ref:`discussed later<how-to:codes:parsing:errors>`).
-This is done through an instance of :py:class:`~aiida.engine.processes.process_spec.CalcJobProcessSpec`, which, as can be seen in the snippet below, is passed as the |spec| argument to the |define| method.
-For the code that adds up two numbers, we will need to define those numbers as inputs (let's call them ``x`` and ``y`` to label them) and the result as an output (``sum``).
-The snippet below shows one potential implementation, as it is included in ``aiida-core``:
+The |define| method tells AiiDA which inputs the |CalcJob| expects and which outputs it produces (exit codes will be :ref:`discussed later<how-to:codes:parsing:errors>`).
+This is done through an instance of the :py:class:`~aiida.engine.processes.process_spec.CalcJobProcessSpec` class, which is passed as the |spec| argument to the |define| method.
+For example:
 
 .. literalinclude:: ../../../aiida/calculations/arithmetic/add.py
     :language: python
     :pyobject: ArithmeticAddCalculation.define
-    :dedent: 4
```

Comment:
Because people will (and should be able to) copy-paste it into their code.
Let's not make life harder unnecessarily

Reply:
Well it makes it harder to read because it is unnecessarily indented. I would argue that is more of a burden than indenting the code once after you have copy pasted the snippet. Anyway, you will have to adapt the code and indenting all code once is a single click in any half-decent editor

Reply:
From my previous tutorial experience, the issue is that not everybody is actually using a decent editor - my main goal here is that I would like to receive zero questions about indentation of this code snippet during the tutorial.

However, you are of course right that indentation ends up hiding a few more characters.
Let's let a 3rd person decide @csadorf What do you say?

Just for reference, here is

NO INDENTATION
![image](https://user-images.githubusercontent.com/1053245/86237600-5c890f80-bb9c-11ea-8552-7199ab3f8729.png)

INDENTATION
![image](https://user-images.githubusercontent.com/1053245/86237744-a07c1480-bb9c-11ea-84dc-df7979b38849.png)

Reply:
I vote against dedenting, because it is more correct and if folks have issues copy & pasting dedented code they will also have issues with figuring out the right detention. But might be worthwhile to record what caused more issues during the actual tutorial.